docs/polkadot/implementors-guide/guide.md

```diff
@@ -0,0 +1,500 @@
+
+[TODO Diagram: Node-side divided into Networking and Core]
+
+Node-side behavior is split up into various Processes. Processes are long-lived workers that perform a particular category of work. Processes can communicate with each other, and typically do so via an Overseer that prevents race conditions.
+
+Runtime logic is divided up into Modules and APIs. Modules encapsulate particular behavior of the system. Modules consist of storage, routines, and entry-points. Routines are invoked by entry points, by other modules, upon block initialization or closing. Routines can read and alter the storage of the module. Entry-points are the means by which new information is introduced to a module. Each block in the blockchain contains a set of Extrinsics. Each extrinsic targets a a specific entry point to trigger and which data should be passed to it. Runtime APIs provide a means for Node-side behavior to extract meaningful information from the state of a fork.
+
+These two aspects of the implementation are heavily dependent on each other. The Runtime depends on Node-side behavior to author blocks, and to include Extrinsics which trigger the correct entry points. The Node-side behavior relies on Runtime APIs to extract information necessary to determine which actions to take.
+
+### Architecture: Node-side
+
+**Design Goals**
+
+* Modularity: Components of the system should be as self-contained as possible. Communication boundaries between components should be well-defined and mockable. This is key to creating testable, easily reviewable code.
+* [TODO] anything else?
+
+[TODO]
+
+---
+
+### Architecture: Runtime
+
+The best architecture at this time is unclear. Let's start by setting down the requirements of the runtime and then trying to come up with an architecture that encompasses all of them.
+
+There are three key points during the execution of a block that we are generally interested in:
+  * initialization: beginning the block and doing set up works. Runtime APIs draw information from the state directly after initialization.
+  * inclusion of new parachain information
+  * finalization: final checks and clean-up work before completing the block.
+
+In order to import parachains, handle misbehavior reports, and keep data accessible, we need to keep this data in the storage/state:
+  * All currently registered parachains.
+  * All currently registered parathreads.
+  * The head of each registered para.
```

Comment:
What is "the head of a para"?

Reply:
We've used the shorthand "para" to mean "parachain or parathread", but I don't know if I've defined this anywhere.

Reply:
Actually my question was more about the "head" part : ) It seems that it is not defined anywhere around

Reply:
oh, by that I meant the current head-data.

Everything I'm writing in this section right now is super messy because I'm just trying to put down all the requirements of the runtime, have others look over them, and then start to propose an architecture that encompasses all of them.

Reply:
single finalised head, or all the unfinalised heads plural?

Reply:
@pepyakin head refers to the tip (most-recent) nodes of a DAG  / history data structure like HEAD in git

Reply:
single finalized head for the view of each relay chain block's state. many unfinalized relay chain heads -> many unfinalized parachain heads.